### PR TITLE
WIP: Add ClassFCCSlotIterator to iterate over FCC in ClassIterator

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
Iterates over classes in the FlattenedClassCache. Called by the
ClassIteratorClassSlots.
Implements:
-nextSlot: Returns next non-NULL class pointer in the FCC
-getIndex: returns the current index of iteration in the FCC

Signed-off-by: Oussama Saoudi <oussama.saoudi@ibm.com>

### CODE REVIEW

File: runtime/gc_structs/ClassFCCSlotIterator.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    UDATA _numberOfFlattenedFields;
+    UDATA _index;
+public:
+	GC_ClassFCCSlotIterator(J9Class *clazz) :
+        _clazz(clazz),
+        _numberOfFlattenedFields(0),
+        _index(0)
+	{
+        if ((J9_ARE_ALL_BITS_SET(_clazz->romClass->modifiers, J9AccValueType)) && (NULL != clazz->flattenedClassCache)) {
+            _numberOfFlattenedFields = clazz->flattenedClassCache->numberOfEntries;
```
Comment (OussamaSaoudi) [inline runtime/gc_structs/ClassFCCSlotIterator.hpp:55]: @dmitripivkine 
Does it matter if I use clazz or _clazz for standard?
Also is the` if((condition1)&&(condition2)) `fine?

Comment (dmitripivkine) [reply runtime/gc_structs/ClassFCCSlotIterator.hpp:55]: 1) yes `clazz`(input parameter) and `_class`(private class member) are equivalent here, you can use any (but the same for consistency) 
2) yes, multiple logical conditions in one if is ok. `J9_ARE_ALL_BITS_SET` is a macro, so you don't need extra round brackets around it. The guide line is put extra round brackets in case there are some operations inside to not rely on order of operations.

**Comment by dmitripivkine** [comment]
What is missed here that Iterator states `classiteratorclassslots_state_*` are used externally, so we should add support for new `classiteratorclassslots_state_flattened_class_cache_slots`:
- in `GC_CheckEngine::checkClassHeap()`: there is trivial handling - just add string description
- in `MM_ReferenceChainWalker::scanClass(): this change requires to introduce new scanning type to`J9GC_REFERENCE_TYPE_*` group. This new type should be added to `mapEventType()` to `jvmtiHeap.c` and `jvmtiHeap10.c`. @tajila Would you please provide instructions how to integrate new class from FCC to JVMTI reporting?

File: runtime/gc_structs/ClassFCCSlotIterator.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * @return the next non-NULL class reference in the flattened class cache
+ * @return NULL if there are no more such references
+ */
+J9Class *
+GC_ClassFCCSlotIterator::nextSlot()
+{
+    J9Class *classPtr = NULL;
+    while (_index < _numberOfFlattenedFields) {
+        classPtr = J9_VM_FCC_CLASS_FROM_ENTRY(J9_VM_FCC_ENTRY_FROM_CLASS(_clazz, _index++));
```
Comment (dmitripivkine) [inline runtime/gc_structs/ClassFCCSlotIterator.cpp:40]: Would you please instead of using auto increment `_index++` change index explicitly in a separate line `_index +=1;` ?

File: runtime/gc_structs/ClassFCCSlotIterator.hpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    J9Class* _clazz;
+    UDATA _numberOfFlattenedFields;
+    UDATA _index;
+public:
+	GC_ClassFCCSlotIterator(J9Class *clazz) :
+        _clazz(clazz),
+        _numberOfFlattenedFields(0),
+        _index(0)
+	{
+        if (J9_ARE_ALL_BITS_SET(_clazz->romClass->modifiers, J9AccValueType) && (NULL != clazz->flattenedClassCache)) {
```
Comment (dmitripivkine) [inline runtime/gc_structs/ClassFCCSlotIterator.hpp:54]: This line is correct, however I am wondering can we save on rom class access...
@tajila Is  check `(NULL != clazz->flattenedClassCache)` sufficient alone or it must be accompanied with check class is VT class?

Comment (OussamaSaoudi) [reply runtime/gc_structs/ClassFCCSlotIterator.hpp:54]: The reason for `J9_ARE_ALL_BITS_SET(_clazz->romClass->modifiers, J9AccValueType)` is that if the value is not a VT, it may have bits in that slot, but not be NULL. 

Ex: This came to light as an issue with array classes, where flattenedClassCache was not null, but it had a value of 0x4. Trying to get an entry from this causes a segfault.

It's worth looking at a more optimal method though.

Comment (dmitripivkine) [reply runtime/gc_structs/ClassFCCSlotIterator.hpp:54]: Proposed code is correct and I like it - it covers exact situation accurately. Reading of rom class also should not be a problem in this case due previous reads in this iterator.
I was wondering this slot is using exclusively for VT classes so check might be not necessary. If it is not a case, so keep existing solution

Comment (tajila) [reply runtime/gc_structs/ClassFCCSlotIterator.hpp:54]: Generally, if you have the ramClass you dont need to fetch to the romClass to check if its a VT, just use `J9_IS_J9CLASS_VALUETYPE(clazz)`.

For this, I think you need to test that its not an array and the FCC is not null. Non-VTs may have classes in the FCC.

```
if (J9_ARE_NO_BITS_SET(J9CLASS_FLAGS(clazz), J9AccClassArray) && (NULL != clazz->flattenedClassCache)) {
```

**Comment by tajila** [comment]
> Would you please provide instructions how to integrate new class from FCC to JVMTI reporting?
 
 We dont have a spec for JVMTI yet, so the exact details on how this should be handled are not yet determined

**Comment by OussamaSaoudi** [comment]
@tajila @dmitripivkine since there is no spec yet, should I leave the section in jvmti blank for this PR?

File: runtime/jvmti/jvmtiHeap.c [hunk 4]
```diff
@@ -1250,6 +1250,7 @@ mapEventType(J9JVMTIHeapData * data, IDATA type, jint index, j9object_t referrer
 		case J9GC_ROOT_TYPE_CLASSLOADER:
 		case J9GC_ROOT_TYPE_JVMTI_TAG_REF:
 		case J9GC_REFERENCE_TYPE_UNKNOWN:
+		case J9GC_REFERENCE_TYPE_FCC:
```
Comment (OussamaSaoudi) [inline runtime/jvmti/jvmtiHeap.c:1253]: @dmitripivkine just to confirm, we want it to fall through and hit the `J9JVMTI_HEAP_EVENT_NONE_NOFOLLOW;` right?

Comment (dmitripivkine) [reply runtime/jvmti/jvmtiHeap.c:1253]: Yes, I think we can do this

File: runtime/gc_include/j9modron.h [hunk 5]
```diff
@@ -145,6 +145,7 @@ typedef jvmtiIterationControl J9MODRON_REFERENCE_CHAIN_WALKER_CALLBACK(J9Object
 #define J9GC_REFERENCE_TYPE_CLASS_ARRAY_CLASS -12 /**< reference to an array class' component type class */
 #define J9GC_REFERENCE_TYPE_CLASS_NAME_STRING -13 /**< reference to a class' name string */
 #define J9GC_REFERENCE_TYPE_CALL_SITE -14 /**< call site reference to an object */
+#define J9GC_REFERENCE_TYPE_FCC -15 /**< reference to a class in flattened class cache */
```
Comment (dmitripivkine) [inline runtime/gc_include/j9modron.h:148]: I would keep explicit notion this is class in the name `J9GC_REFERENCE_TYPE_CLASS_FCC`

File: runtime/gc_structs/ClassIteratorClassSlots.cpp [hunk 6]
```diff
@@ -77,7 +77,14 @@ GC_ClassIteratorClassSlots::nextSlot()
...
 		_state += 1;
-		
+
+	case classiteratorclassslots_state_flattened_class_cache_slots:
+		classPtr = _classFCCSlotIterator.nextSlot();
+		if (NULL != classPtr) {
+			return classPtr;
+		}
+		_state += 1;
+
```
Comment (dmitripivkine) [inline runtime/gc_structs/ClassIteratorClassSlots.cpp:87]: Would you please add formal state `classiteratorclassslots_state_end:` here (just before `default:`)?

File: runtime/gc_include/j9modron.h (hunk 5 above)
Comment (OussamaSaoudi) [reply runtime/gc_include/j9modron.h:148]: Does `J9GC_REFERENCE_TYPE_FCC_CLASS` work fine?

File: runtime/gc_include/j9modron.h [hunk 7]
```diff
@@ -145,6 +145,7 @@ typedef jvmtiIterationControl J9MODRON_REFERENCE_CHAIN_WALKER_CALLBACK(J9Object
 #define J9GC_REFERENCE_TYPE_CLASS_ARRAY_CLASS -12 /**< reference to an array class' component type class */
 #define J9GC_REFERENCE_TYPE_CLASS_NAME_STRING -13 /**< reference to a class' name string */
 #define J9GC_REFERENCE_TYPE_CALL_SITE -14 /**< call site reference to an object */
+#define J9GC_REFERENCE_TYPE_FCC_CLASS -15 /**< reference to a class in flattened class cache */
```
Comment (dmitripivkine) [inline runtime/gc_include/j9modron.h:148]: I am sorry, I think `J9GC_REFERENCE_TYPE_CLASS_FCC` is better for consistency with `J9GC_REFERENCE_TYPE_CLASS_ARRAY_CLASS` and `J9GC_REFERENCE_TYPE_CALL_SITE`

File: runtime/gc_structs/ClassFCCSlotIterator.hpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * @file
+ * @ingroup GC_Structs
+ */
+
+#if !defined(CLASSFCCSLOTITERATOR_HPP_)
+#define CLASSFCCSLOTITERATOR_HPP_
+
+#include "j9.h"
```
Comment (dmitripivkine) [inline runtime/gc_structs/ClassFCCSlotIterator.hpp:32]: I think it is better to include `j9nonbuilder.h` explicitly as far as this file contain structure and macro definitions for FCC

**Review by dmitripivkine — APPROVED** [review]
Looks good to me

**Comment by dmitripivkine** [comment]
@tajila Would you please review? I believe it is ready. Also would you please suggest proper testing to have something in FCC?

**Comment by tajila** [comment]
Jenkins test sanity,extended xlinuxval jdknext

**Comment by tajila** [comment]
Jenkins test sanity win jdk8

**Comment by tajila** [comment]
Jenkins test sanity xlinuxvalst jdknext
